modyn/trainer_server/internal/trainer/remote_downsamplers/remote_craig_downsampling.py

```diff
@@ -0,0 +1,196 @@
+            raise ValueError(
+                f"The required Greedy optimizer is not available. Pick one of the following: {OPTIMIZER_CHOICES}"
+            )
+
+        # This class uses the embedding recorder
+        self.requires_coreset_supporting_module = True
+
+        # Samples are supplied label by label (this class is instance of AbstractPerLabelRemoteDownsamplingStrategy).
+        # The following list keeps the gradients of the current label. When all the samples belonging to the current
+        # label have been seen, the scores are computed and the list is emptied
+        self.current_class_gradients: list[torch.Tensor] = []
+        # distance_matrix[i,j] = 0 if label[i]!=label[j] else is proportional to the euclidean distance between the
+        # two samples in the gradient space
+        self.distance_matrix = np.zeros([0, 0])
+
+        # if true, the downsampling is balanced across classes ex class sizes = [10, 50, 30] and 50% downsampling
+        # yields the following downsampled class sizes [5, 25, 15] while without balance something like [0, 45, 0] can
+        # happen
+        self.balance = params_from_selector["balance"]
+        if self.balance:
+            self.already_selected_samples: list[int] = []
+            self.already_selected_weights = torch.tensor([]).float()
+
+    def inform_samples(
+        self,
+        sample_ids: list[int],
+        forward_output: torch.Tensor,
+        target: torch.Tensor,
+        embedding: Optional[torch.Tensor] = None,
+    ) -> None:
+        assert embedding is not None
+
+        # Slightly different implementation for BTS and STB since in STB points are supplied class by class while in
+        # BTS are not. STB will always use the first branch, STB will typically (might use the first if all the points
+        # belong to the same class) use the second one
+        different_targets_in_this_batch = target.unique()
+        if len(different_targets_in_this_batch) == 1:
+            self._inform_samples_single_class(sample_ids, forward_output, target, embedding)
+        else:
+            for current_target in different_targets_in_this_batch:
+                mask = target == current_target
+                this_target_sample_ids = [sample_ids[i] for i, keep in enumerate(mask) if keep]
+                self._inform_samples_single_class(
+                    this_target_sample_ids, forward_output[mask], target[mask], embedding[mask]
+                )
+                self.inform_end_of_current_label()
+
+    def _inform_samples_single_class(
+        self, sample_ids: list[int], forward_output: torch.Tensor, target: torch.Tensor, embedding: torch.Tensor
+    ) -> None:
+        embedding_dim = embedding.shape[1]
+        num_classes = forward_output.shape[1]
+        batch_num = target.shape[0]
+
+        loss = self.criterion(forward_output, target).mean()
+
+        # compute the gradient for each element provided
+        with torch.no_grad():
```

Comment:
I may be stupid, but how can we compute gradients with torch.no_grad?

Reply:
`Torch.no_grad()` is applied to the operations within the context manager (matrix multiplication) and not to the previous ones (forward pass). Hence, we can still compute the gradient we need, and we avoid computing the gradient of the product (kind of gradient of the gradient), which would be useless.

Reply:
sorry, I was braindead here apparently